### PR TITLE
Add Activity Log Serializer

### LINKED ISSUE #125: Add Serializer for Activity Log
## Overview

Add a serializer for the Activity Log, and include its data in the `/boundaries/{id}/` endpoint.

It should be a list of activities, ordered most recent first. Each activity should have an action ("Approved" "Reviewed" "Submitted" "Drafted"), an actor (user's name), and a date. Some activities may also have notes. Reviews will also have annotations.

### PR DESCRIPTION
**Pull request by rachelekm**
## Overview

Adds activity log data as part of the `/boundaries/{id}/` endpoint. Each activity in the log should include the activity type, user taking the action, and time of action as well as any notes and, if an activity is submitting a review, number of annotations. Activity types should be “Drafted", "Submitted", "Reviewed", "Approved".

Closes #125 

### Demo

Screen Shot 2022-10-14 at 6 42 45 PM

## Testing Instructions

I updated the test data to include a handful more submissions to better test the ability to order log events. The below endpoint includes multiple submissions:
http://localhost:8181/api/boundaries/4/

- `scripts/resetdb`
- Login and visit the above api endpoints
- Confirm the activity log is returned as expected
- Login to admin and create a couple various submission activities for a boundary
- Confirm activity log updates correctly and in order for that boundary detail endpoint
- `scripts/test` to confirm no errors

## Checklist

- [x] `fixup!` commits have been squashed
- [x] `CHANGELOG.md` updated with summary of features or fixes, following [Keep a Changelog](https://keepachangelog.com/en/1.0.0/) guidelines
- [x] `README.md` updated if necessary to reflect the changes
- [x] CI passes after rebase

### CODE REVIEW

File: src/django/api/tests.py
Comment (jacobtylerwalls) [inline src/django/api/tests.py:81]: At your option, as `update` feels more of a bulk-operation thing.
```suggestion
        draft.created_at = datetime(2022, 10, 2, hour=9, tzinfo=timezone.utc)
        draft.save()
```

File: src/django/api/tests.py
Comment (jacobtylerwalls) [inline src/django/api/tests.py:142]: Not related to your changes, but found in testing. Would you mind adding `reviewed_at` to the the review object, a few lines above? We may not need both `created_at` and `reviewed_at`, but that's another discussion.

File: src/django/api/tests.py
Comment (jacobtylerwalls) [inline src/django/api/tests.py:173]: IMO we shouldn't overwrite the `approved` variable name, which can get confusing! (I see this was already happening with some names here.)

File: src/django/api/serializers/activity_log.py
Comment (jacobtylerwalls) [inline src/django/api/serializers/activity_log.py:48]: I noticed reviews and approvals are blocking activity entries for submissions/drafts. For instance, if you visit http://localhost:8181/api/boundaries/5/, you get a single entry for an approval. I would have expected to have also seen the drafted, submitted, etc.

We can exchange some solutions over Slack?

File: src/django/api/views/boundary.py
Comment (jacobtylerwalls) [inline src/django/api/views/boundary.py:58]: Purely cosmetic AFAICT, but you can just `select_related('review', 'approval')`

File: CHANGELOG.md
Comment (jacobtylerwalls) [inline CHANGELOG.md:45]:
```suggestion
- Add Activity Log Serializer [#140](https://github.com/azavea/iow-boundary-tool/pull/140)
```

**Review by jacobtylerwalls — CHANGES_REQUESTED** [review]
I just had one question about capturing the original drafted/submitted actions after a review or approval of a boundary. The rest is cosmetic feedback at your option.

File: src/django/api/tests.py
Comment (jacobtylerwalls) [reply src/django/api/tests.py:81]: At some future point we might create a datetime helper that could take just a number of hours after 10/1/2022 and produce the wanted datetimes.

Comment (rajadain) [reply src/django/api/tests.py:81]: We're also specifying a value for `created_at` in the creation of all other instances here. Do they also not work correctly? Or can we just use do that?

Comment (rajadain) [reply src/django/api/tests.py:81]: Perhaps something like this:

```python
def dt(date: str, hour: int = 0) -> datetime:
    """
    Turns a string of type "YYYY-MM-DD" into a datetime.
    Uses the optional hour value if specified, else defaults to 0.
    """
    year = int(date[:4])
    month = int(date[5:7])
    day = int(date[8:])

    return datetime(year, month, day, hour, tzinfo=timezone.utc)
```
```python
   created_at=dt('2022-10-03', hour=15),
```

**Review by rajadain — COMMENTED** [review]
Same here, that I'm not getting the full activity log for the approved `boundaries/5`. Should see when it was first drafted, submitted, reviewed, drafted again, submitted again, and approved.

File: src/django/api/tests.py
Comment (jacobtylerwalls) [reply src/django/api/tests.py:81]: > We're also specifying a value for created_at in the creation of all other instances here. Do they also not work correctly? Or can we just use do that?

I did it that way to reduce the variation between test runs. If it became important to structure test cases so that action B took place in between actions A and C, I didn't want to rely on the microsecond timing between parts of this test method executing.

I might be missing your point, though!

Comment (rajadain) [reply src/django/api/tests.py:81]: > I might be missing your point, though!

My point is specifying `created_at` in

```python
Submission.objects.create(created_at=...)
```

vs changing the value after creation:

```python
s = Submission.objects.create(...)
s.created_at = ...
s.save()
```

Does that make sense?

Comment (jacobtylerwalls) [reply src/django/api/tests.py:81]: Ah, yes it does!

Comment (rajadain) [reply src/django/api/tests.py:81]: So should we change it everywhere then? Not just here?

Comment (rajadain) [reply src/django/api/tests.py:81]: Alternatively, if we just created the items in the right order, that should be enough, and we may drop `created_at=` specification completely.

Comment (rachelekm) [reply src/django/api/tests.py:81]: Ah, good call. The latter option seems like it would be more intuitive/readable, so I'm for that! However I understand Jacob's point on reducing potential variations. Maybe a compromise is dropping the specification to improve readability and adding a test to check that the order of execution of submissions creation is correct?

Comment (jacobtylerwalls) [reply src/django/api/tests.py:81]: I retract my concern :-)

For now I bet a comment is fine, we can always add the test later.

Comment (rachelekm) [reply src/django/api/tests.py:81]: Thanks all, will do a comment! If there's time an activity log test could potentially be dual-purpose and confirm serializer output as well as the correct order, too.

File: src/django/api/tests.py
Comment (rachelekm) [reply src/django/api/tests.py:173]: Whoops, good catch!

File: src/django/api/serializers/activity_log.py
Comment (rachelekm) [reply src/django/api/serializers/activity_log.py:26]: oh that's helpful clarification, thank you!

File: src/django/api/tests.py
Comment (rachelekm) [reply src/django/api/tests.py]: Since we are adding a "Review Started" activity log for now, that object works well as a "review started but not yet submitted" test, so I left it as is for now.

**Comment by rachelekm** [comment]
Thanks, all! Edits are in and ready for another round of reviews. As discussed on slack I added in "Review Started" as another action to log—I left off any notes/annotations for those as it seemed like something only worth displaying once the review is submitted.

File: src/django/api/serializers/activity_log.py
Comment (rajadain) [inline src/django/api/serializers/activity_log.py:59]: Approvals should not have any notes:

image

File: src/django/api/serializers/activity_log.py
Comment (rajadain) [inline src/django/api/serializers/activity_log.py:39]: Seeing the contributor as the user for Review activities: http://localhost:8181/api/boundaries/4/

image

It should be the validator.

File: src/django/api/serializers/activity_log.py
Comment (rajadain) [inline src/django/api/serializers/activity_log.py:21]: Drafts should not have notes:

image

File: src/django/api/tests.py
Comment (rajadain) [inline src/django/api/tests.py:67]:
```suggestion
        # So there is potential for order variation if unexpected execution sequence
```

**Review by rajadain — CHANGES_REQUESTED** [review]
Code organization looks good! Some inaccuracies in the output though.

**Comment by rachelekm** [comment]
Thanks for catching those! Just made the `notes` field optional too since there's no reason to send back an empty string anymore.

File: src/django/api/serializers/activity_log.py
Comment (rajadain) [inline src/django/api/serializers/activity_log.py:32]: This can be simplified to

```suggestion
        obj['user'] = instance.submitted_by.email
```

Don't need to

```python
submitter = User.objects.get(pk=instance.submitted_by.pk)
```

File: src/django/api/tests.py
Comment (rajadain) [inline src/django/api/tests.py:67]:
```suggestion
        # So there is potential for order variation if unexpected execution sequence
```

**Review by rajadain — APPROVED** [review]
+1 nice work on this! Left a couple minor code comments, thence this is good to merge.

File: src/django/api/tests.py
Comment (rachelekm) [reply src/django/api/tests.py:67]: Only minorly embarrassed this got me twice 🤦‍♀️

**Review by jacobtylerwalls — APPROVED** [review]
Looks great!

File: src/django/api/tests.py
Comment (rachelekm) [reply src/django/api/tests.py:108]: Good catch! That'll be a nice cleanup. I'll leave it for now as you mentioned since we'll go back over these in the next couple cards.

**Pull request merged**
